pwnlib/libcdb.py

```diff
@@ -97,30 +104,94 @@ def provider_libc_rip(hex_encoded_id, hash_type):
 
     if not data:
         log.warn_once("Could not fetch libc binary for %s %s from libc.rip", hash_type, hex_encoded_id)
+        return None, ""
+
+    return data, libc_match[0]['id']
+
+# Search offline https://github.com/niklasb/libc-database for symbols
+def find_local_libc(params):
+    if not context.local_libcdb:
+        log.warn_once("Before enable offline search, please configure `PWNLIB_LOCAL_LIBCDB` or `context.local_libcdb`.")
+        return None
+
+    if not params.get("symbols"):
         return None
-    return data
 
-PROVIDERS = [provider_libcdb, provider_libc_rip]
+    res = []
+    db_path = Path(context.local_libcdb) / "db"
+
+    for symbols_path in db_path.rglob("*.symbols"):
+        syms = _get_libc_symbols(symbols_path)
+
+        matched = 0
+        for name, addr in params["symbols"].items():
+            if isinstance(addr, str):
+                addr = int(addr, 16) 
+
+            if syms.get(name) and syms.get(name) & 0xfff == addr & 0xfff:
+                matched += 1
+
+        if matched == len(params["symbols"]):
+            libs_id = symbols_path.stem
+            libc_path = os.path.join(db_path, "%s.so" % libs_id)
+            libs_url = read(os.path.join(db_path, "%s.url" % libs_id)).decode().strip()
+            res.append(_pack_libs_info(libc_path, libs_id, libs_url, syms))
+
+    return res
+
+# Offline search https://github.com/niklasb/libc-database for hash type
+def offline_provider_libc_database(hex_encoded_id, hash_type):
+    if not context.local_libcdb:
+        log.warn_once("Before enable offline search, please configure `PWNLIB_LOCAL_LIBCDB` or `context.local_libcdb`.")
+        return None, ""
+
+    db_path = Path(context.local_libcdb) / "db"
+    hashfilehex = None
+
+    if hash_type == "build_id":
+        hash_type = "buildid"
 
-def search_by_hash(hex_encoded_id, hash_type='build_id', unstrip=True):
+    # "buildid" is not in HASHES, so we can correct calculate file hash
+    if hash_type in HASHES:
+        hashfilehex = eval("%sfilehex" % hash_type)
```

Comment:
HASHES is just a string list.
```
HASHES = ['build_id', 'sha1', 'sha256', 'md5']
```

The `eval` is only used to dynamically fetch the hash function based on the `hash_type` value, such as md5, sha1, and so on. This enables us to calculate the file's hash and compare it to the `hex_encoded_id`.

Reply:
`eval` function is very error-prone (and quite slow anyway); if any of the surrounding semantics change, many things can go wrong. Please change `HASHES` to a mapping like `{"sha1": sha1filehex, ...}` and just index it. This will stay in line with 'the Zen of Python': `explicit is better than implicit`.

Note that if you ever find yourself in need of global variable lookup by dynamic name, you can use [`globals()["do_" + whatever]`](https://docs.python.org/3/library/functions.html#globals) as a last resort. `eval` is almost never a good idea.

Reply:
I will try to address it.